doc/differentiation/autodiff.qbk

```diff
@@ -0,0 +1,805 @@
+    using variable = typename nested_dimensions<RealType,Order,Orders...>::type;
+    
+    // Satisfies Boost's Conceptual Requirements for Real Number Types.
+    // Don't use this dimension<> class directly. Instead, use the variable<> alias.
+    template<typename RealType,size_t Order>
+    class dimension
+    {
+      public:
+    
+        // Initialize a variable of differentiation.
+        explicit dimension(const root_type&);
+    
+        // The root data type chosen when variable<> is declared. E.g. float, double, etc.
+        explicit operator root_type() const;
+    
+        // Query return value of computation to get the derivatives.
+        template<typename... Orders>
+        typename type_at<RealType,sizeof...(Orders)-1>::type derivative(Orders... orders) const;
+    
+        // All of the arithmetic operators are overloaded.
+        template<typename RealType2,size_t Order2>
+        dimension<RealType,Order>& operator+=(const dimension<RealType2,Order2>&);
+    
+        dimension<RealType,Order>& operator+=(const root_type&);
+    
+        // ...
+    };
+    
+    // Standard math functions are overloaded and called via argument-dependent lookup (ADL).
+    template<typename RealType,size_t Order>
+    dimension<RealType,Order> floor(const dimension<RealType,Order>&);
+    
+    template<typename RealType,size_t Order>
+    dimension<RealType,Order> exp(const dimension<RealType,Order>&);
+    
+    // ...
+    
+    } } } } // namespace boost::math::differentiation::autodiff
+    
+
+[h1:description 
+Description]
+
+Autodiff is a header-only C++ library that facilitates the [@https://en.wikipedia.org/wiki/Automatic_differentiation automatic differentiation] (forward mode) of mathematical functions in single and multiple variables.
+
+The formula central to this implementation of automatic differentiation is the following [@https://en.wikipedia.org/wiki/Taylor_series Taylor series] expansion of an analytic function [autodiff_equation form_0.png] at the point [autodiff_equation form_1.png]: 
+
+[:[:[autodiff_equation form_2.png]]]
+
+
+The essential idea of autodiff is the substitution of numbers with polynomials in the evaluation of [autodiff_equation form_0.png]. By selecting the proper polynomial [autodiff_equation form_3.png] as input, the resulting polynomial contains the function's derivatives within the polynomial coefficients. One then multiplies by a factorial term to obtain the desired derivative of any order.
+
+Assume one is interested in the first [autodiff_equation form_4.png] derivatives of [autodiff_equation form_0.png] at [autodiff_equation form_1.png]. Then without any loss of precision to the calculation of the derivatives, all terms [autodiff_equation form_5.png] that include powers of [autodiff_equation form_6.png] greater than [autodiff_equation form_4.png] can be discarded, and under these truncation rules, [autodiff_equation form_0.png] provides a polynomial-to-polynomial transformation: 
```

Comment:
If the equations are trivial, then +1 for writing them in markup rather than png's.
Otherwise how are they authored - Tex?  Can you generate SVG's for those - I don't recall if that's possible from Tex or not?

Reply:
MathJax generates SVGs; I've explicitly extracted it out of the DOM before, but it ain't pretty.

Reply:
I really want to include the mathematics in the documentation. In response to a later [question](https://github.com/boostorg/math/pull/176#discussion_r247605420) about just referring to a reference, I derived the mathematics independently. For instance I think the wikipedia entry on AD is too confusing and obscures the explanation using the chain rule. Focusing on polynomial arithmetic as it relates to the Taylor series is the simplest way to explain this IMHO. Thus I wanted to have the source documentation be in MathJax, which makes doxygen a natural choice. I wrote a separate script (html2qkb.rb) that translated the html doxygen output into QuickBook, and the png images that result are presented nicely when I am viewing them. Here is a screenshot:

![image](https://user-images.githubusercontent.com/39707/51137692-266f9880-180d-11e9-9f41-f615e450a845.png)

I guess you're saying that it's not going to look as nicely for all users?

I am hesitant to make autodiff.qbk the primary source of documentation, as it doesn't handle equations well. To do so would be quite a loss, IMO. Hope this makes sense, and perhaps an alternate suggestion can be made that takes this into consideration, if possible.

Reply:
I am able to reproduce the out-of-line pngs if the /doc/src/boostbook.css is not included. This can happen if an http server is started in a subdirectory of /libs/math. However if the root of the web service is the root boost directory, then the inclusion of this file will allow the following css rule to be applied to the inline png images

    span.inlinemediaobject img
    {
        vertical-align: middle;
    }

which fixes the vertical alignment issue you are seeing. I imagine this is the way people will generally read the documentation.

Reply:
> I'd definitely say keep the display-style equations as SVG

What's wrong with PNG, as you originally suggested? I'd be happy to if there was an easy way that doxygen could output SVG equations. I did look into that originally and didn't see an easy way, but I can try again if I could see the motivation.

> But it will be very hard to maintain a pile of .pngs whose contents are just an italic letter.

Will it help if I open-source the doxygen html-to-qkb converter?

Sorry this is not ideal, but it is also not ideal to use a format like QuickBook for a mathematics library that doesn't support rendering mathematics.

Reply:
The display-style equations as PNG are ok. But I think the single character pngs are a step too far.

Reply:
> In terms of references, what do you think is a good one to refer readers to? I'd still prefer to see at least one reference in the docs.

My primary/only source was the [wikipedia](https://en.wikipedia.org/wiki/Automatic_differentiation) page. I'll add that as a reference.

> I think the single character pngs are a step too far.

I have certainly considered this. For example, this is precisely what I did on the README page on the front of https://github.com/pulver/autodiff since that web page does not have the benefit of the boostbook.css file to fix the vertical alignment of the inline images. So in this case, the italiced versions are preferred. The drawback to do this with the autodiff.qbk file however is 2-fold:

1. It requires manual edits to the output of html2qbk.rb file. This would be repeated each time significant changes were made to the documentation.
2. It looks worse than the above screenshot.

It would help me to see your perspective if you could give a use-case/scenario that would make it worth these two drawbacks.

Reply:
Once the file goes into boost, the .qbk needs to be the primary point of authority for the documentation, since John has to maintain it for years. So the custom tool `html2.qbk.rb` cannot be part of the system.

Reply:
+1 to that: for better or worse we're using quickbook, and we need to keep all the Math lib docs in the one format otherwise it becomes unmaintainable down the road.

We do desperately need a better way to handle equations though...  I do notice that there are online tools that seem to convert LaTex to svg really rather well: http://www.tlhiv.org/ltxpreview/.  But of course it's clunky to have to switch to a separate tool just for equation authoring.  There are a bunch of suggestions for batch/command line conversions here: https://askubuntu.com/questions/33196/how-to-convert-latex-equations-to-svg and another more recent script here: https://www.npmjs.com/package/tex-equation-to-svg.

Assuming they work OK, one quick-ish solution would be to have a quickbook template:

[LaTex filename LaTex goes here^2]

which would simply expand to:

[$filename.svg]

And then run a short script over the quickbook files to extract and write the .svg's

Thoughts?

Reply:
How's this: The boost autodiff.qbk documentation will be signicantly trimmed down to focus on C++ usage, with a minimal amount of mathematics. I'll keep the mathematical intro, which will involve two SVG equations to provide the basic outline of the theory. All other inline math symbols will be italicezed as Nick suggests. A [link](http://www.unitytechgroup.com/doc/autodiff/) to my more in-depth documentation will be provided which uses MathJax for the notation.

Reply:
@pulver : I think this is a good solution. Another very good option for you is since you believe the Wikipedia to be incomplete, is just for you to edit wikipedia. I will be doing this anyway once your feature lands (otherwise no one seems to find what we have), so it will be nice to have an internally consistent wiki page.

Reply:
There are now 7 svg equations in `autodiff.qbk` and the minor inline symbols are non-images as suggested.

Also the external link to more in-depth documentation is replaced with a link to an autodiff.pdf that comes with the library, as John suggested below. (I'll continue to improve this documentation, but I'm going to very quickly have less time to work on this project so I hope we can push out what is here and make improvements to the documentation in the next/later versions.)